src/OmniSharp.Roslyn.CSharp/Services/IntelliSense/V2/CompletionService.cs

```diff
@@ -2,68 +2,72 @@
+    using CompletionModels = OmniSharp.Models.V2.Completion;
 
+    [Shared]
     [OmniSharpHandler(OmniSharpEndpoints.V2.Completion, LanguageNames.CSharp)]
-    internal class CompletionService : IRequestHandler<Models.CompletionRequest, Models.CompletionResponse>
+    [OmniSharpHandler(OmniSharpEndpoints.V2.CompletionItemResolve, LanguageNames.CSharp)]
+    internal class CompletionService :
+        IRequestHandler<CompletionModels.CompletionRequest, CompletionModels.CompletionResponse>,
+        IRequestHandler<CompletionModels.CompletionItemResolveRequest, CompletionModels.CompletionItemResolveResponse>
     {
         private readonly OmniSharpWorkspace _workspace;
         private readonly ILogger _logger;
 
+        private (string fileName, CompletionList completionList) _lastResponse;
+
         [ImportingConstructor]
         public CompletionService(OmniSharpWorkspace workspace, ILoggerFactory loggerFactory)
         {
             _workspace = workspace;
             _logger = loggerFactory.CreateLogger<CompletionService>();
         }
 
-        public async Task<Models.CompletionResponse> Handle(Models.CompletionRequest request)
+        public async Task<CompletionModels.CompletionResponse> Handle(CompletionModels.CompletionRequest request)
         {
             var fileName = request.FileName;
             var position = request.Position;
             var trigger = request.Trigger.ToRoslynCompletionTrigger();
 
-            var document = _workspace.GetDocument(fileName);
-            if (document == null)
-            {
-                throw new ArgumentException($"Could not find document for {fileName}.");
-            }
+            var document = GetDocument(fileName);
 
             var text = await document.GetTextAsync();
             if (position < 0 || position > text.Length)
             {
-                throw new ArgumentOutOfRangeException($"Invalid position: {position}. Should be with range 0 to {text.Length}");
+                throw new ArgumentOutOfRangeException($"Invalid position: {position}. Should be within range 0 to {text.Length}");
             }
 
-            var service = CSharpCompletionService.GetService(document);
-            if (service == null)
-            {
-                throw new InvalidOperationException("Could not retrieve Roslyn CompletionService.");
-            }
+            var service = GetService(document);
 
             if (!service.ShouldTriggerCompletion(text, position, trigger))
             {
-                return Models.CompletionResponse.Empty;
+                return CompletionModels.CompletionResponse.Empty;
             }
 
             var completionList = await service.GetCompletionsAsync(document, position, trigger);
+            if (completionList == null)
+            {
+                return CompletionModels.CompletionResponse.Empty;
+            }
 
             var isSuggestionMode = completionList.SuggestionModeItem != null;
```

Comment:
This loses the placeholder text for the item, and I don't think there is a way to resolve the description text for the builder anymore.